apps/server/src/modules/tool/uc/tool-reference.uc.ts

```diff
@@ -0,0 +1,100 @@
+import { ForbiddenException, Injectable } from '@nestjs/common';
+import { ToolReference } from '@shared/domain/domainobject/tool/tool-reference';
+import {
+	ContextExternalToolDO,
+	ContextRef,
+	EntityId,
+	ExternalToolDO,
+	Permission,
+	SchoolExternalToolDO,
+	ToolConfigurationStatus,
+} from '@shared/domain';
+import { Action } from '@src/modules/authorization';
+import {
+	CommonToolService,
+	ContextExternalToolService,
+	ExternalToolService,
+	SchoolExternalToolService,
+} from '../service';
+import { ToolContextType } from '../interface';
+import { ToolReferenceMapper } from '../controller/mapper/tool-reference.mapper';
+
+@Injectable()
+export class ToolReferenceUc {
+	constructor(
+		private readonly externalToolService: ExternalToolService,
+		private readonly schoolExternalToolService: SchoolExternalToolService,
+		private readonly contextExternalToolService: ContextExternalToolService,
+		private readonly commonToolService: CommonToolService
+	) {}
+
+	async getToolReferences(userId: EntityId, contextType: ToolContextType, contextId: string): Promise<ToolReference[]> {
+		const contextRef = new ContextRef({ type: contextType, id: contextId });
+
+		const contextExternalTools: ContextExternalToolDO[] = await this.contextExternalToolService.findAllByContext(
+			contextRef
+		);
+
+		const toolReferencesPromises: Promise<ToolReference | null>[] = contextExternalTools.map(
+			(contextExternalTool: ContextExternalToolDO) => this.buildToolReference(userId, contextExternalTool)
+		);
+
+		const toolReferencesWithNull: (ToolReference | null)[] = await Promise.all(toolReferencesPromises);
+		const filteredToolReferences: ToolReference[] = toolReferencesWithNull.filter(
+			(toolReference: ToolReference | null): toolReference is ToolReference => toolReference !== null
+		);
+
+		return filteredToolReferences;
+	}
+
+	private async buildToolReference(
+		userId: EntityId,
+		contextExternalTool: ContextExternalToolDO
+	): Promise<ToolReference | null> {
+		try {
+			await this.ensureToolPermissions(userId, contextExternalTool);
+		} catch (e: unknown) {
+			if (e instanceof ForbiddenException) {
```

Comment:
This looks wrong and is maybe also toggled with the other comment from me.

Reply:
For the behavior it is correct, because we want to filter out tools which the user is not allowed to use, because he is on another school for example. Maybe he copied a launch link from a user from another school.
 
In the future we could also have shared courses/boards/learnrooms and so on..